# Exploring Ebay Car Sells Data Set

## 1. Introduction

This is project is part of the Data Quest Data Cleaning and Visualization mission and aims to clean a web scrapped data set available in [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data) using Pandas and the built in vectorized
functions.

In order to analyze the data set, the NumPy and Pandas modules are imported to be open and read the data set in a suitable format (csv).

### 1.1 Importing modules and creating a panda dataframe

In [1]:
import pandas as pd
import numpy as np
#Reading the csv data set into panda dataframe
autos = pd.read_csv('autos.csv',encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

As `heads()` and `info()` methods shows, the columns of autos data frame comprises two types of objects: *int 64* and *object*.

Some features are numerical data types but are in a string-like format like **odometer** and **price**. The name is in non-uniform written format varying from row to row what is expected require some hard work to clear them. 

### 1.2. Modyfing column names in data set

In [4]:
columns = autos.columns
print(columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


The column names are written in CamelCase which consists in joining the upper case words of a name. We will change the column names to snakecase which join lower case words and separate by underscore (eg. vehicle_type instead of vehicleType). 

This is only a matter of personal preferece but can get confusing for long column names. So, the CamelCase names will be changed to SnakeCase names.

In [5]:
name_snake = {'yearOfRegistration':'registration_year', 'monthOfRegistration':'registration_month','notRepairedDamage':'unrepaired_damage','dateCreated':'ad_created'}
autos.rename(name_snake,axis=1,inplace=True)
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


### 1.3. Investigating and cleaning time related variables

In [6]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuelType  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [7]:
#Descriptive statistics about autos dataframe
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  registration_year  gearbox       powerPS  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuelType       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

The description of the data set shows some columns with values that need to be corrected. the **odometer** and **price** columns for instance are non-null object types, but have to be corrected to float64 type.

The postal code is a int64 object but it should be an object because it has no mathematical meaning, only refers to some adress.

Some registration years, the maximum and minimum specifically, show non meaningful values such as 9999 or 1000. This column needs more investigation.



In [8]:
#cleaning price column
#Investigatin unique values 
print(autos["price"].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [9]:
# Removing '$' and ',' string and coverting to float
autos["price"] = autos["price"].str.replace('$','')
autos["price"] = autos["price"].str.replace(',','')
autos["price"] = autos["price"].astype(float)

In [10]:
autos['price'][0:3]

0    5000.0
1    8500.0
2    8990.0
Name: price, dtype: float64

In [11]:
# Removing 'km' and ',' from odometer column
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(float)

In [12]:
# renaming odometer column
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)
autos['odometer_km'][0:3]

0    150000.0
1    150000.0
2     70000.0
Name: odometer_km, dtype: float64

In [13]:
#odometer and price statistics
print(autos['odometer_km'].describe())
print(autos['price'].describe())


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [14]:
# Odometer value counts
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_km, dtype: int64

The price statistics shows unrealistic values while the odometer seems to have reasonable values. The minimum value is zero for instance. In practical terms, it would mean a free car. However, it does not make any sense to anounce a car for free in a web page. 

Also the maximum value is 1E+8 dolars which is a totally out of range figure for a car price. It is worth investigating the car brand and type for that price.

In [15]:
# Checking 100 milion dolar car brand and type
milioncar_brand = autos.loc[autos['price'] == 1e+8,'brand']
milioncar_type = autos.loc[autos['price']==1e+8,'vehicleType']
freecar_brand = autos.loc[autos['price'] == 0,'brand'].describe()
freecar_type = autos.loc[autos['price']==0,'vehicleType'].value_counts()


print(freecar_brand)
print(freecar_type[0:4])
print(milioncar_brand)
print(milioncar_type)



count           1421
unique            38
top       volkswagen
freq             348
Name: brand, dtype: object
limousine     256
kleinwagen    237
kombi         194
coupe          71
Name: vehicleType, dtype: int64
Series([], Name: brand, dtype: object)
Series([], Name: vehicleType, dtype: object)


The analysis shows that the 100 milion car has actually no brand and Type. Then, it is a good decision to remove it from data set.

It can also be seen that lots of cars have zero price which actually mean they are for free (yey!). But it does not make any sense to announce a car for free. However, I will keep it since it accounts for lots of data.

In [16]:
new_autos=autos[(autos['price']> 0) & (autos['price']<1e+8)].head() 
#it does not make sense to drop zero price. we get with about 5 values

In [17]:
new_autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5 entries, 0 to 4
Data columns (total 20 columns):
dateCrawled           5 non-null object
name                  5 non-null object
seller                5 non-null object
offerType             5 non-null object
price                 5 non-null float64
abtest                5 non-null object
vehicleType           5 non-null object
registration_year     5 non-null int64
gearbox               5 non-null object
powerPS               5 non-null int64
model                 5 non-null object
odometer_km           5 non-null float64
registration_month    5 non-null int64
fuelType              5 non-null object
brand                 5 non-null object
unrepaired_damage     5 non-null object
ad_created            5 non-null object
nrOfPictures          5 non-null int64
postalCode            5 non-null int64
lastSeen              5 non-null object
dtypes: float64(2), int64(5), object(13)
memory usage: 840.0+ bytes


Now it's time to work with the columns embedded by the crawler. They are 'date_crawled', 'last seen'. We will also work with 'ad_created' column, which was not crawled.

In [18]:
autos[['dateCrawled','ad_created','lastSeen']][0:5]

dateCrawled           ad_created             lastSeen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [19]:
# frequency table of months for ad created
year_create = autos['ad_created'].str[0:4]
freq_year_create = year_create.value_counts(normalize=True,dropna=False)
freq_year_create.sort_index(ascending=True)
print(freq_year_create)

2016    0.99988
2015    0.00012
Name: ad_created, dtype: float64


it looks like people from this sample like to sell cars in march and april.

Another variable of interest is the year of registration it is also numeric and time related. The `series.describe()` method will be used to investigate it.

In [20]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The maximum and minimum values for this series are out of a reasonable range for a car registration year. Let's count how many cases of outliers (higher than 2016 or lower than 1900) there is. A car cannot be registered after it was announced for sell. Thus the upper limit will be set to 2016 which is the highest year for created ads.

In [21]:
out_of_range = autos.loc[(autos["registration_year"] > 2016)|(autos["registration_year"] < 1900),"registration_year"].value_counts(dropna=False)
print(out_of_range)
print('the total number of outliers are',out_of_range.sum())

2017    1453
2018     492
9999       4
5000       4
2019       3
9000       2
1800       2
6200       1
4500       1
8888       1
4800       1
2800       1
1001       1
1000       1
1111       1
1500       1
9996       1
5911       1
4100       1
Name: registration_year, dtype: int64
the total number of outliers are 1972


### 1.4 Investigating the brand column

In this section, the variables related to brand will be analyzed. the criteria to choose which brand is included is the relative frequence it appears in the data set. Brands below 5% are not included.

In [51]:
average_price = {}
freq_brands=autos["brand"].value_counts(dropna=False,normalize=True)
freq_brands_abv5 = freq_brands[freq_brands >= 0.05]
print(freq_brands_abv5.index)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [83]:
for row in freq_brands_abv5.index:
    brand = row
    avg_price = autos.loc[autos["brand"]==brand,"price"].mean()
    average_price[brand]=avg_price
panda_average_price = pd.Series(average_price,name='avg_price_$')
print(panda_average_price.sort_values(ascending=False))

mercedes_benz    29511.955429
audi              8965.560355
bmw               8252.918954
ford              7105.662547
volkswagen        6384.167400
opel              5106.092657
Name: avg_price_$, dtype: float64


The analysis shows that **Mercedes_Benz** is the car brand with the highest average price (about 30k) between the ones with relative frequence greater than 5% in the data set. Its average price is about 4 times higher than the audi and bmw (8.5k) which are in 2nd and 3rd place in the rank. The brand with lowest average price is opel (5k).

let's try finding a justification for that with the mileage variable. If the mileage of merceds bens cars are lower on average, it is a good indication that it is a contributing factor for the higher average prices and not only the brand positioning in the market.

In [87]:
average_miles = {}
for row in freq_brands_abv5.index:
    brand = row
    avg_mileage = autos.loc[autos["brand"]==brand,"odometer_km"].mean()
    average_miles[brand]=avg_mileage
panda_avg_mile = pd.Series(average_miles,name='avg_miles_km')
df = pd.concat([panda_avg_mile,panda_average_price],axis=1)
df

avg_miles_km   avg_price_$
audi           129643.941163   8965.560355
bmw            132521.643028   8252.918954
ford           124131.934464   7105.662547
mercedes_benz  130886.142797  29511.955429
opel           129298.663248   5106.092657
volkswagen     128955.272761   6384.167400

The new aggregated data frame shows that the average miles of merceds bens cars are not higher on average than the other brands. It reflects the brand market positioning showing that its selling value is significant higher than the other brands for the same usage conditions.

The registration time could also be used as an indication of usage intensity to correlate with average price thought it is certainly correlated with mileage. 